src/broker/index.js

```diff
@@ -341,20 +341,23 @@ module.exports = class Broker {
   /**
    * @public
    * @param {number} replicaId=-1 Broker id of the follower. For normal consumers, use -1
-   * @param {object} topics e.g:
-   *                  [
-   *                    {
-   *                      topic: 'topic-name',
-   *                      partitions: [
-   *                        { partition: 0 }
-   *                      ]
-   *                    }
-   *                  ]
+   * @param {number} isolationLevel=1 This setting controls the visibility of transactional records (default READ_COMMITTED, Kafka >0.11 only)
+   * @param {TopicPartitionOffset[]} topics e.g:
+   *
+   * @typedef {Object} TopicPartitionOffset
+   * @property {string} topic
+   * @property {PartitionOffset[]} partitions
+   *
+   * @typedef {Object} PartitionOffset
+   * @property {number} partition
+   * @property {number} [timestamp=-1]
+   *
+   *
    * @returns {Promise}
    */
-  async listOffsets({ replicaId, topics }) {
+  async listOffsets({ replicaId, isolationLevel, topics }) {
     const listOffsets = this.lookupRequest(apiKeys.ListOffsets, requests.ListOffsets)
-    return await this.connection.send(listOffsets({ replicaId, topics }))
+    return await this.connection.send(listOffsets({ replicaId, isolationLevel, topics }))
```

Comment:
I would suggest doing the normalization between versions here, rather than in the cluster. We don't really have an established pattern for where to normalize. We have done it in the protocol in some place (unfortunately I can't remember where...), but I think we should try to keep our protocol true to the Kafka protocol. But we shouldn't let that bleed through too far, like the cluster, so I think the broker is the best place for it that we have at the moment.

Reply:
Ok. I _do_ like currently that the broker is a thin wrapper to execute the protocols. I also agree that we should not handle the normalization in the protocols (it's done for fetch with record sets vs messages, and it's giving me some trouble now that we're working with transactions). I think ideally there'd be a wrapper around the broker that did normalization. But I'm fine moving into the broker for now, just didn't want to introduce a new pattern without consensus 😄